packages/react/src/Banner/Banner.tsx

```diff
@@ -0,0 +1,323 @@
+import cx from 'clsx'
+import React, {createContext, useContext, useEffect, useId, useMemo} from 'react'
+import styled from 'styled-components'
+import {AlertIcon, InfoIcon, StopIcon, CheckCircleIcon, XIcon} from '@primer/octicons-react'
+import {Button, IconButton} from '../Button'
+import {get} from '../constants'
+
+type BannerVariant = 'critical' | 'info' | 'success' | 'upsell' | 'warning'
+
+export type BannerProps = React.ComponentPropsWithoutRef<'section'> & {
+  /**
+   * Provide an optional description for the Banner. This should provide
+   * supplemental information about the Banner
+   */
+  description?: React.ReactNode
+
+  /**
+   * Provide an icon for the banner.
+   * Note: Only `variant="info"` banners should use custom icons
+   */
+  icon?: React.ReactNode
+
+  /**
+   * Optionally provide a handler to be called when the banner is dismissed.
```

Comment:
It looks like we're expecting the consumer to manage focus upon dismiss and that worries me a little bit given it seems very easy to forget.

I'm curious to hear your thoughts on:

* why we would or wouldn't want to build focus on dismiss into the component.
* how we might encourage people to implement focus management correctly.

I'd love if we can leverage some of the programmatic focus management logic that Lindsey introduced to the global flash. 🤔

Reply:
@khiga8 great points, thanks for bringing them up. Maybe a small change we could make here to accommodate what your suggesting is to make is so that anytime `onDismiss` is provided that we have an explicit `returnFocusRef` (or something similar) prop so that focus is always explicitly managed? We could also make it so that `onDismiss` expects a return value which is the element to return focus to.

Happy to explore either option, with potentially the latter feeling more similar to what I'm assuming people will write (similar to the storybook example).

Let me know what you think! Happy to bake in this, or other approaches, into the component 👍

Reply:
> We could also make it so that onDismiss expects a return value which is the element to return focus to.

Are you thinking that the component would then handle placing focus on whatever the return value is?

Reply:
> Are you thinking that the component would then handle placing focus on whatever the return value is?

Exactly 👍 

We could also totally follow this behavior when `onDismiss` is called (specifically the `getElementToFocus` helper) and restore it that way, as well. Would it be worth making an issue for this for us to explore/try out different implementations or is this something that should be incorporated in this PR?

Reply:
> Would it be worth making an issue for this for us to explore/try out different implementations

This sounds good to me! :)

Reply:
Added over at: https://github.com/github/primer/issues/3222

Would love to pair on this if you're interested! Also wanted to check to see if it would be okay if this work was in a separate PR?

Reply:
Definitely down to pair! 

My understanding is that this is a draft/first implementation and something that will be iterated on based on learnings :)